test/frontend-unit-tests/spec/controllers/login.js

```diff
@@ -7,40 +7,72 @@ describe('Controller: LoginCtrl', function () {
 
   var LoginCtrl,
     scope,
-    $httpBackend;
+    rootScope,
+    UserService,
+    loginDeferred,
+    location,
+    loginForm;
 
   // Initialize the controller and a mock scope
-  beforeEach(inject(function (_$httpBackend_, $controller, $rootScope) {
-    $httpBackend = _$httpBackend_;
-    scope = $rootScope.$new();
-    LoginCtrl = $controller('LoginCtrl', {
-      $scope: scope
-    });
+  beforeEach(inject(function ($controller, _$rootScope_, $q) {
+    rootScope = _$rootScope_;
+    scope = rootScope.$new();
+
+    loginDeferred = $q.defer();
+    location = jasmine.createSpyObj('location', ['path']);
+    UserService = jasmine.createSpyObj('UserService', ['login']);
+    UserService.login.and.returnValue(loginDeferred.promise);
 
-    // mock angular form
-    scope.optionsForm = {
-      model: {},
-      email: {
-        $error: {}
+    rootScope.setCurrentUser = jasmine.createSpy('setCurrentUser');
+
+    loginForm = {
+      username: {
+        $error: {
+          required: false
+        }
       }
     };
-    scope.optionsForm.model.$setValidity = function () {};
 
-    // mock user
-    scope.user = {
-      email: '',
-      password: '',
-      username: ''
-    };
+    LoginCtrl = $controller('LoginCtrl', {
+      $scope: scope,
+      $rootScope: rootScope,
+      $location: location,
+      userService: UserService
+    });
+
+    scope.user = {};
```

Comment:
It shouldn't be there. Now we handle the validation on the view side, allowing the user to login only when the user and password are complete.